ddtrace/internal/module.py

```diff
@@ -0,0 +1,427 @@
+                module_name = module.__spec__.name
+                module_path_base = abspath(join(*module_name.split(".")))
+
+                def _test_py_file(base_path):
+                    if isfile(base_path + ".py"):
+                        return base_path + ".py"
+                    return None
+
+                # Check if it is a .py file
+                module_origin = _test_py_file(module_path_base)
+                if module_origin is None:
+                    # If not, check if it is a __main__.py file.
+                    module_origin = _test_py_file(join(module_path_base, "__main__"))
+
+                if module_origin is not None:
+                    # Register the module origin mapping
+                    sys.modules._origin_map[module_origin] = module  # type: ignore[attr-defined]
+
+                    # Temporarily override the __file__ attribute to allow
+                    # any registered hooks to run.
+                    old_file = module.__file__
+                    module.__file__ = module_origin
+                    try:
+                        sys.modules.after_import(module)  # type: ignore[attr-defined]
+                    finally:
+                        module.__file__ = old_file
+
+            def _(*args, **kwargs):
+                abstract_code = Bytecode.from_code(args[0])
+                for i, instr in enumerate(abstract_code):
+                    if not isinstance(instr, Instr):
+                        continue
+
+                    if instr.name == "MAKE_FUNCTION" and abstract_code[i + 1].name == "STORE_NAME":
+                        # DEV: This is a bit dumb. Every time we find a function
+                        # definition, we re-trigger the hooks to see if there is
+                        # a new function to patch. This code is unlikely to be
+                        # exercised in practice, but it is important to be aware
+                        # of this behaviour.
+                        lineno = abstract_code[i + 1].lineno
+                        abstract_code[i + 2 : i + 2] = Bytecode(
+                            [
+                                Instr("LOAD_CONST", _run_module_hook, lineno=lineno),
+                                Instr("CALL_FUNCTION", 0, lineno=lineno),
+                                Instr("POP_TOP", lineno=lineno),
+                            ]
+                        )
+
+                args[1][_run_module_hook.__name__] = _run_module_hook
+                args = (abstract_code.to_code(),) + args[1:]
+                return cls._run_code(*args, **kwargs)
+
+            runpy._run_code = _  # type: ignore[attr-defined]
+
+    @classmethod
+    def is_installed(cls):
+        """Check whether this module watchdog class is installed."""
+        return any(type(_) is cls for _ in sys.meta_path)
+
+    @classmethod
+    def uninstall(cls):
+        # type: () -> None
+        """Uninstall the module watchdog."""
```

Comment:
Probably worth noting in here that multiple patches will be uninstalled?

Reply:
This will remove just the most recent one. Agreed to improve the docstring to better describe this behaviour.